### PR TITLE
First draft

### PR DESCRIPTION
**Pull request by TomArcher**

### CODE REVIEW

**Comment by TomArcher** [comment]
@danroth27 @Erikre @Rick-Anderson @rynowak Please review

File: mvc/controllers/areas.rst [hunk 1]
```diff
@@ -1,8 +1,132 @@
-.. include:: /../common/stub-topic.txt
+Areas
+=====
+By `Tom Archer`
 
-|stub-icon| Areas
-===========================================
+MVC uses a folder hierarchy pattern that defines the relationship between your controllers, views, and actions. You can see this with the default wizard-generated ASP.NET MVC app where you start with a ``Controllers`` folder containing several controllers (``Account``, ``Home``, and ``Manage``). If you open the ``HomeController.cs`` file, you'll note that the ``HomeController`` controller has three actions - ``Index``, ``About``, and ``Contact`` - that return a view object without qualifying which view to return. (A fourth action - ``Error`` - is present, but it explicitly names the location of the view to return.) So, how does MVC know which view to return when its location is not explicitly stated?
```
Comment (Rick-Anderson) [inline mvc/controllers/areas.rst]: ``HomeController.cs`` is not code, it's a file so use *file*

Comment (Rick-Anderson) [inline mvc/controllers/areas.rst]: we don't have editors so lets do it right - drop the redundant ``HomeController`` controller  or write it as Home controller.

File: mvc/controllers/areas.rst [hunk 2]
```diff
@@ -1,8 +1,132 @@
...
 
-.. include:: /../common/stub-notice.txt
+MVC finds the associated view for a controller action by searching in two folders: ``Views/[controller][action].cshtml`` and ``Views/Shared/[action].cshtml``. Therefore, when the ``HomeController.About`` method attempts to create a view without specifying the location of the view, the MVC routing engine will look for ``Views/Home/About.cshtml`` first. If that's not found, the routing engine will look for ``Views/Shared/About.cshtml``. If neither are found, an exception is thrown.
```
Comment (Rick-Anderson) [inline mvc/controllers/areas.rst]: It doesn't create a view, find the correct word. It's also not rendering a view. You should be able to look up the right word to use in our MSDN docs

Comment (Rick-Anderson) [inline mvc/controllers/areas.rst]: You need to generalize this. You are assuming return View(); with no params. What happens when you add a param ("About")? Same search.

Comment (MikeWasson) [inline mvc/controllers/areas.rst]: [controller][action] > [controller]/[action]

File: mvc/controllers/areas.rst [hunk 3]
```diff
@@ -1,8 +1,132 @@
...
 
-.. _issue: https://github.com/aspnet/Docs/issues/121
+This pattern works well for many Web apps. However, for larger apps, this folder hierarchy can become difficult to manage as you have controllers in one parent folder, and views in a separate parent folder. This is where `Areas` come in. Areas enable you to partition your project such that controllers and their views can be stored under a parent folder that defines their logical grouping.
+
+Let's take a look at an example to illustrate how Areas are created and used. In this example, we'll take a simple store app (MyStore) that has a several logical groupings of controllers and actions. These groupings are ``Services``, ``Products``, and ``Deals``. The following steps will walk you through creating an area for each grouping, as well as the controllers and views for each.
+
+#. In Solution Explorer, right-click the project name and create a new folder named ``Areas``.
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: This topic seems aimed at users who already have a basic "getting started" knowledge of MVC, so I'd consider skipping the detailed, UI-based steps here. Just show the folder hierarchy and the code.

File: mvc/controllers/areas.rst [hunk 4]
```diff
@@ -1,8 +1,132 @@
...
+
+Once you've defined the folder hierarchy, you need to tell MVC that each controller is associated with an area. You do that by decorating the controller name with the ``[Area]`` attribute. Here's what the boiler-plate ``Products/HomeController`` looks like with the attribute added.
+
+.. code-block:: c#
+  :emphasize-lines: 4
+
+  ...
+  namespace MyStore.Areas.Products.Controllers
+  {
+      [Area("Products")]
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: Does the area name here need to match the area folder name? (I guess my specific question is: Does routing happen based on the name in the attribute, or the folder name?)

File: mvc/controllers/areas.rst [hunk 5]
```diff
@@ -1,8 +1,132 @@
...
+
+.. code-block:: c#
+  :emphasize-lines: 4-6
+
+  ...
+  app.UseMvc(routes =>
+  {
+      routes.MapRoute(name: "areaRoute",
+              template: "{area:exists}/{controller}/{action}",
+              defaults: new { controller = "Home", action = "Index" });
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: What is the difference between `defaults: new {  }` and `{controller=Home}` below?

File: mvc/controllers/areas.rst [hunk 6]
```diff
@@ -1,8 +1,132 @@
...
+
+      routes.MapRoute(
+          name: "default",
+          template: "{controller=Home}/{action=Index}");
+  });
+
+Now, when the user enters http://<yourApp>/products, the Index action method of the ``HomeController`` in the ``Products`` area will be invoked and its view will be sent to the client for rendering.
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: Not sure what "user enters" means here ... say "a client requests" or something?

File: mvc/controllers/areas.rst (hunk 4 above)
Comment (TomArcher) [inline mvc/controllers/areas.rst]: @MikeWasson Yes, they need to match.

File: mvc/controllers/areas.rst (hunk 6 above)
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: The view is rendered on the server side.

Comment (TomArcher) [inline mvc/controllers/areas.rst]: I meant "browses to". I'll fix it.

File: mvc/controllers/areas.rst [hunk 7]
```diff
@@ -1,8 +1,132 @@
...
+    - HomeController
+
+  - Views
+
+    - Home
+
+      - Index
+
+As you can see, this example has two ``HomeController`` controllers; one that's defined in an area (``Products/Controllers/HomeController``) and one that's defined outside an area (``Controllers/HomeController``). Let's say you want to put a link on each page that links back to the other page.
+
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: How do you link across areas? (I'm guessing the "outside to inside" syntax.)

File: mvc/controllers/areas.rst [hunk 8]
```diff
@@ -1,8 +1,132 @@
...
+Linking from inside to outside
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Linking from inside to outside simply means to link from a place that's inside an area to an action that's outside an area. As you can probably guess, you simply specify an area of "".
+
+.. code-block:: c#
+
+  @Html.ActionLink("Go to Home Page", "Index", "Home", new { area = "" }, null)
+
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: I feel like this whole section would be clearer if it just said: You disambiguate controllers by including the area in the route values. Use an empty string for controllers that aren't within an area.  And then include the two code examples.

File: mvc/controllers/areas.rst (hunk 6 above)
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: Also maybe call out explicitly that the URL path for an action under an area is `/area/controller/action`. That's implicit in the route template but might not be obvious.

File: mvc/controllers/areas.rst (hunk 4 above)
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: Maybe call that out. (Aside... from a quick test, it seems *routing* is by the Area attribute name, and view lookup is by folder name. If they don't match, you'll get a "view not found" error.)

File: mvc/controllers/areas.rst (hunk 1 above)
Comment (danroth27) [inline mvc/controllers/areas.rst]: > default wizard-generated ASP.NET MVC app 
default ASP.NET MVC app template

File: mvc/controllers/areas.rst (hunk 2 above)
Comment (danroth27) [inline mvc/controllers/areas.rst]: I think these first two paragraphs should just be removed. There isn't anything about areas in these first two paragraphs. The default MVC conventions should be documented elsewhere.

File: mvc/controllers/areas.rst [hunk 9]
```diff
@@ -1,8 +1,132 @@
...
 
-.. include:: /../common/stub-notice.txt
+MVC finds the associated view for a controller action by searching in two folders: ``Views/[controller][action].cshtml`` and ``Views/Shared/[action].cshtml``. Therefore, when the ``HomeController.About`` method attempts to create a view without specifying the location of the view, the MVC routing engine will look for ``Views/Home/About.cshtml`` first. If that's not found, the routing engine will look for ``Views/Shared/About.cshtml``. If neither are found, an exception is thrown.
 
-.. _issue: https://github.com/aspnet/Docs/issues/121
+This pattern works well for many Web apps. However, for larger apps, this folder hierarchy can become difficult to manage as you have controllers in one parent folder, and views in a separate parent folder. This is where `Areas` come in. Areas enable you to partition your project such that controllers and their views can be stored under a parent folder that defines their logical grouping.
```
Comment (danroth27) [inline mvc/controllers/areas.rst]: > as you have controllers in one parent folder, and views in a separate parent folder.

This isn't really the reason things get difficult to manage. The reason is that as apps get bigger you end up with a lot of controllers and views, which makes the app difficult to reason about. Areas let you partition your app into logic chunks that are smaller and easier to manage. It's this logical partitioning that is important, not the folder structure (note that the folder structure really only matters for views in MVC 6).

File: mvc/controllers/areas.rst (hunk 3 above)
Comment (danroth27) [inline mvc/controllers/areas.rst]: +1. Remove the point and click instructions and instead focus on documenting the runtime behavior.

File: mvc/controllers/areas.rst (hunk 4 above)
Comment (danroth27) [inline mvc/controllers/areas.rst]: Right, putting the code for the controllers in a folder along side the corresponding views is a good practice, but it is not required by the runtime. It's the `[Area]` attribute on the controller that determines which area it is in.

File: mvc/controllers/areas.rst (hunk 5 above)
Comment (danroth27) [inline mvc/controllers/areas.rst]: Please use the inline template syntax instead of the anonymous object: `{area:exists}/{controller=Home}/{action=Index}`

**Comment by TomArcher** [comment]
@danroth27 @Rick-Anderson @MikeWasson @Erikre @TomArcher Please review

File: mvc/controllers/areas.rst [hunk 10]
```diff
@@ -1,8 +1,93 @@
...
+Areas
+=====
+By `Tom Archer`
 
-|stub-icon| Areas
-===========================================
+:doc:`Areas </controllers/areas>` provide a way to separate a large MVC application into semantically-related groups of models, views, and controllers. Let's take a look at an example to illustrate how Areas are created and used. Let's say you have a store app that has two distinct groupings of controllers and views: Products and Services.
 
-.. include:: /../common/stub-notice.txt
+Instead of having all of the controllers housed under the Controllers parent directory, and all the views housed under the Views parent directory, you could use Areas to group your views and controllers according to the area (or logical grouping) with which they're associated.
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: nit: "housed" sounds odd to me. "located"?

File: mvc/controllers/areas.rst [hunk 11]
```diff
@@ -1,8 +1,93 @@
...
-.. _issue: https://github.com/aspnet/Docs/issues/121
+- Project name
+
+  - Areas
+
+    - Products
+
+      - Controllers
+
+        - HomeController
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: HomeController.cs

File: mvc/controllers/areas.rst [hunk 12]
```diff
@@ -1,8 +1,93 @@
...
+      {
+          // GET: /<controller>/
+          public IActionResult Index()
+          {
+              return View();
+          }
+      }
+  }
+
+The final step is to set up a route definition that works with your newly created areas. The :doc:`routing` article goes into detail about how to create route definitions, including using conventional routes versus attribute routes. In this example, we'll use a conventional route. To do so, simply open the ``startup.cs`` file and modify it by adding the highlighted route definition below.
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: Startup.cs (initial cap) - to match the project templates

File: mvc/controllers/areas.rst [hunk 13]
```diff
@@ -1,8 +1,93 @@
...
+        name: "default",
+        template: "{controller=Home}/{action=Index}");
+  });
+
+Now, when the user browses to *http://<yourApp>/products*, the ``Index`` action method of the ``HomeController`` in the ``Products`` area will be invoked.
+
+Linking between areas
+---------------------
+
+To link between areas, you simply specify the area in which the controller is defined. If the controller doesn't exist within an area, use an empty string.
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: Maybe say "If the controller is not part of an area"? ("Doesn't exist..." makes it sound like an error condition...)

File: mvc/controllers/areas.rst [hunk 14]
```diff
@@ -1,8 +1,93 @@
...
+
+The following snippet shows how to link from to a controller action that is defined within an area named *Products*.
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: "from to" ... missing a word here?

File: mvc/controllers/areas.rst [hunk 15]
```diff
@@ -1,8 +1,93 @@
...
+        - HomeController
+
+      - Views
+
+        - Home
+
+          - Index.cshtml
+
+Once you've defined the folder hierarchy, you need to tell MVC that each controller is associated with an area. You do that by decorating the controller name with the ``[Area]`` attribute.
+
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: Call out that the views are resolved using the file path? i.e., folder naming is significant. You need an "Areas" folder etc. `[area]/Views/[controller]/[action].cshtml` ... The example shows what to do, but it's not immediately clear to me that names are significant.

File: mvc/controllers/areas.rst [hunk 16]
```diff
@@ -1,8 +1,93 @@
...
+  {
+    routes.MapRoute(name: "areaRoute",
+      template: "{area:exists}/{controller=Home}/{action=Index}");
+
+    routes.MapRoute(
+        name: "default",
+        template: "{controller=Home}/{action=Index}");
+  });
+
+Now, when the user browses to *http://<yourApp>/products*, the ``Index`` action method of the ``HomeController`` in the ``Products`` area will be invoked.
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: Should there be a section on using shared views in areas? Looks like you can use either `/Areas/[area]/Views/Shared/` or global `/Views/Shared`

File: mvc/controllers/areas.rst [hunk 17]
```diff
@@ -1,8 +1,93 @@
...
+
+To link to a controller action that is outside an area, simply specify an empty string.
+
+.. code-block:: c#
+
+  @Html.ActionLink("Go to Home Page", "Index", "Home", new { area = "" }, null)
+
+Summary
+-------
+Areas are a very useful tool for grouping semantically-related controllers and actions under a common parent folder. In this article, you learned how to set up your folder hierarchy to support ``Areas``, how to specify the ``[Area]`` attribute to denote a controller as belonging to a specified area, and how to define your routes with areas. Finally, you saw how to link between areas - including linking from outside to inside, and linking from inside to outside.
```
Comment (MikeWasson) [inline mvc/controllers/areas.rst]: With this revision, you got rid of the "outside to inside" terminology in the previous section (which I approve :smiley:) so maybe delete the last phrase, "including....."

File: mvc/controllers/areas.rst (hunk 16 above)
Comment (TomArcher) [inline mvc/controllers/areas.rst]: Fixed

**Comment by Erikre** [comment]
:shipit:

**Pull request merged**